rfcs/20200710-reduction/README.md

```diff
@@ -0,0 +1,159 @@
+  where eps_mode = {add, max}
+* TensorFlow:
+  L2_normalize(x) = x[:] / pow(max(sum(pow(x[:], 2)), eps), 0.5)
+    = x[:] / pow(max((pow(norm_l2(x[:]), 2)), eps), 0.5)
+* PyTorch:
+  L2_normalize(x) = x[:] / max(pow(sum(pow(x[:], 2)), 0.5), eps)
+    = x[:] / max(norm_l2(x[:]), eps)
+
+# Proposal
+## Option 1: Implement reduction primitive and norm primitive.
+
+### Motivation
+Motivation for this option is to keep reduction primitive API simple, but limit
+flexibility and amount of supported algorithms.
+
+### API
+
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_norm_l1,
+    dnnl_reduction_norm_l2,
+    dnnl_reduction_norm_linf,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    dnnl_reduction_mean,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    // Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
+        dnnl_reduction_desc_t *reduction_desc,
+        dnnl_alg_kind_t reduction_alg_kind,
+        const dnnl_memory_desc_t *src_desc,
+        const dnnl_memory_desc_t *dst_desc);
+```
+
+## Option 2: Implement reduction primitive and extend eltwise post operations.
+
+### Motivation
+Motivation for this option is to support more algorithms extending reduction
+primitive using eltwise algorithm and multiple post operations. However this
+option complicates reduction primitive API. Also it complicates implementations
+because of multiple post operations which should be fused in an efficient way
+to be on par with monolithic implementation.
+
+### Details
+
+norm_l2 = sqrt(reduction(pow, sum, 2.0, 0, x))
+
+### API
+
+``` cpp
+// dnnl_types.h
+typedef enum {
+    ...
+    dnnl_eltwise_abs,
+    dnnl_eltwise_pow,
+    ...
+    dnnl_reduction_max,
+    dnnl_reduction_min,
+    dnnl_reduction_sum,
+    dnnl_reduction_mul,
+    ...
+} dnnl_alg_kind_t;
+
+typedef struct {
+    dnnl_primitive_kind_t primitive_kind;
+    // The kind of eltwise algorithm applied before reduction. This
+    // algorithm can be undefined.
+    dnnl_alg_kind_t eltwise_alg_kind;
+    // The kind of reduction algorithm.
+    dnnl_alg_kind_t reduction_alg_kind;
+    // Source memory descriptor.
+    dnnl_memory_desc_t src_desc;
+    // Destination memory descriptor.
+    dnnl_memory_desc_t dst_desc;
+} dnnl_reduction_desc_t;
+
+// dnnl.h
+dnnl_status_t DNNL_API dnnl_reduction_desc_init(
+        dnnl_reduction_desc_t *reduction_desc,
+        dnnl_alg_kind_t eltwise_alg_kind,
+        dnnl_alg_kind_t reduction_alg_kind,
+        float alpha, float beta,
+        const dnnl_memory_desc_t *src_desc,
+        const dnnl_memory_desc_t *dst_desc);
```

Comment:
not sure I understand how this API maps on to the operation mentioned above:

> `norm_l2 = sqrt(reduction(pow, sum, 2.0, 0, x))`

Reply:
norm_l2 = sqrt(reduction(pow, sum, 2.0, 0, x)) = reduction(/*eltwise alg */ alg:pow, /* reduction alg */ alg:sum, /* eltwise alpha */ 2.0, /* eltwise beta */ 0, x) + POST_OPS(eltwise(alg:pow, 0.5))

Reply:
Indeed :)